### ISSUE TITLE
Nested Rules

### ISSUE THREAD

**Issue by Haravikk** [JohnDoh/Roundcube-Plugin-SieveRules-Managesieve#118]
It occurs to me that my usage of the advanced editor for sieve rules isn't actually all that advanced; i.e- the only reason I need to use it is because I needed to add a nested conditional to one of the my filters, which currently isn't possible under the standard interface.

I'd like to propose that nesting and `if`/`elsif`/`else` style logic be exposed, as this will allow more complex rules to be defined without having to switch to the advanced editor.

In essence this could be achieved by adding a new action type "Filter", which spawns the same basic interface again, allowing nested conditions to be defined. The only difference is that additional filters can be added to the end to add `elsif` conditions, with the option of defining a final "all messages" (`else`) filter. So long as the interface is collapsable it shouldn't become too overwhelming, but one option would be to allow the sub-filters to be named so they can be presented as a hierarchy in the filters list.

Not a simple feature, but as with many rules features in mail programs the simple editor is just a little too simple IMO. Fine for very basic rules, but its limitations quickly make themselves felt, however nested filters would cover many of these cases without having to resort to the advanced editor.

Of course I'm deeply thankful for that advanced editor option, as in mail clients like Apple's Mail.app on OS X we're not so fortunate, and managing rules in it sucks as a result ;)

**Comment by JohnDoh** [comment]
-1 from me. i guess in certain advanced situations that might be useful, but as you said you can use the advanced interface for that.

**Comment by Haravikk** [comment]
The problem really is that this means we have the situation where the simple interface is a bit too simple, and the advanced editor is way over most user's heads. The question really is how complex it would be to actually handle the nesting; the interface is essentially there already, and the code generation is basically the same as well.

Thinking about it more, the simplest way to handle the UI would be to allow filters to be nested in the filters list, e.g- the same way as managing folders. The `if`/`elsif`/`else` can be done by position in the list (first sub-filter is an `if`, the rest are `elsif` unless they match everything). Selecting a sub-filter would just give exactly the interface there is now for regular (top level) filters.

The only other question mark is how to handle actions added to the parent filter; logically they would go at the top (before sub-filters are processed) but they might be more useful at the bottom for actions in common.

If you're not interested in it, I do actually have experience developing in PHP, so I might have a look at hacking a copy of the plugin that I currently have installed, so if you have any idea of where I'm likely to run into problem that would be a huge help!

**Comment by JohnDoh** [comment]
I think you are trying to solve a problem which does not exist.

**Comment by Kofl** [comment]
Just my 2 cents: For the mass of users, its fine how it is at the moment.

**Comment by IzzySoft** [comment]
@JohnDoh 

> I think you are trying to solve a problem which does not exist.

Well, it exists for me too. Or is there an easy way I've overlooked to have a rule like `if (mail from X && (subject contains A || subject contains B || subject equals C))`? True, this rather "simple example" could be solved using a regex match. But `if (mail from X && (subject contains A || subject contains B || header flag C is 'foo')` could not. Yes, one could setup multiple rules to work around that – but a) this would be less performant and b) prolong the "list of rules" unnecessarily (plus making it harder to manage if it one day gets quite long ;)

I recently added such a "nested rule" manually to my sieve file. Result: I no longer can use your plugin to manage my rules while not having SSH access to the mail server. Plus, for simple things, I still prefer the GUI :) Apart from the fact that I couldn't locate the Advanced Editor in the WebUI (at the moment, I cannot check the config as I'm at a customer's site w/o SSH access to anywhere outside the company).

I agree it's nothing urgent/prio-1, and I also agree the majority of the users won't miss it. Still, it would be really nice to have, if possible.
